Zadanie102

jtermi opened the pull request.

jtermi committed:
zadanie wykonane

jtermi committed:
zadanie wykonane

slawekludwiczak commented on src/main/java/pl/javastart/task/Subscription.java:
```diff
@@ -0,0 +1,47 @@
+
+public class Subscription extends Agreement {
+
+    public Subscription(double monthlyPayment) {
+        super();
+        this.monthlyPayment = monthlyPayment;
+    }
+
+    private double monthlyPayment;
```
Kolejność kodu:
- najpierw pola
- później konstruktory
- dalej pozostałe metody

slawekludwiczak commented on src/main/java/pl/javastart/task/Agreement.java:
```diff
@@ -0,0 +1,60 @@
+    private int smsCount = 0;
+    private int mmsCount = 0;
+    private int secCount = 0;
+    private double accountState = 0;
```
Czy to jest na pewno wspólne dla wszystkich umów?
W Subscription to pole zawsze będzie miało wartość 0

slawekludwiczak commented on src/main/java/pl/javastart/task/Prepaid.java:
```diff
@@ -0,0 +1,68 @@
+            this.setSmsCount(this.getSmsCount() + 1);
+            setAccountState(getAccountState() - smsPrice);
+            System.out.println("SMS wysłany\n");
+            return;
+        }
+        System.out.println("Nie udało się wysłać SMSa\n");
```
Zamiast returna bardziej czytelne jest użycie else

slawekludwiczak commented on src/main/java/pl/javastart/task/Prepaid.java:
```diff
@@ -0,0 +1,68 @@
+            this.setMmsCount(this.getMmsCount() + 1);
+            setAccountState(getAccountState() - mmsPrice);
+            System.out.println("MMS wysłany\n");
+            return;
+        }
+        System.out.println("Nie udało się wysłać MMSa\n");
```
Zamiast returna bardziej czytelne jest użycie else

slawekludwiczak commented on src/main/java/pl/javastart/task/Prepaid.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    public void makePhoneCall(int callDuration) {
+
+        double accountstate = this.getAccountState();
```
Kolejne wyrazy wielkimi literami, czyli `accountState`

slawekludwiczak commented on src/main/java/pl/javastart/task/Mix.java:
```diff
@@ -0,0 +1,122 @@
+package pl.javastart.task;
+
+public class Mix extends Agreement {
```
Ta umowa ma wiele wspólnego z Prepaidem, więc lepiej dziedziczyć właśnie po tamtej klasie

slawekludwiczak commented on src/main/java/pl/javastart/task/Mix.java:
```diff
@@ -0,0 +1,122 @@
+        this.perMinPrice = perMinPrice;
+    }
+
+    @Override
+    public void sendSms() {
+
+        int smsCount = getSmsCount();
+        if (smsLimit > 0) {
+            System.out.println("SMS wysłany\n");
+            smsLimit--;
+            setSmsCount(smsCount + 1);
+            return;
+        }
+        double accountState = getAccountState();
+        if (accountState > smsPrice) {
+            System.out.println("SMS wysłany\n");
+            setSmsCount(smsCount + 1);
+            setAccountState(accountState - smsPrice);
+            return;
+        }
+        System.out.println("Brak środków na koncie\n");
+    }
+
```
Jeżeli będziesz dziedziczył po klasie Prepaid, to zarówno w tej metodzie, jak i pozostałych, będziesz mógł wykorzystać część logiki z klasy Prepaid, wywołując metody przez `super.nazwaMetody()`

slawekludwiczak reviewed (changes requested):
Przydałoby się wydzielić osobne pakiety, żeby jasno wskazać co możemy robić, a czego nie. Przykładowo w main chcemy:
- tworzyć umowy
- tworzyć telefon
- wysyłać sms/mms, lub dzwonić z użyciem telefonu
Ale w metodzie main niekoniecznie chcemy dawać możliwość bezpośredniej manipulacji umową, np. 
```
Prepaid prepaid = new Prepaid(2, 0.1, .2, .5);
prepaid.setSmsCount(50);
```
Jeżeli przeniesiesz wszystkie klasy, oprócz Main, do osobnego pakietu, to możesz np. zrezygnować z getterów i setterów, a zamiast tego użyć specyfikatora domyślnego

jtermi committed:
wprowadzone poprawki:
- nazwa jednego z pól nie zgodna z konwencją
- korzystanie z agregacji
- wyodrębnienie wspólnego kodu do osobnych metod
- przeniesienie klas związanych z umowami do osobnego pakietu i skorzystanie ze specyfikatorów domyslnych wewnątrz nowego pakietu
- usunięcie geterów i seterów
- przeniesienie pola accountState do Prepaid
- zmiana dziedziczenia klas (Mix teraz dziedziczy po Prepaid)
- zamiana "return;" wykorzystywanego do wyjścia z IF{} na na ELSE {}

slawekludwiczak commented on src/main/java/pl/javastart/task/internal/Agreement.java:
```diff
@@ -0,0 +1,27 @@
+    public Agreement() {
+    }
+
+    public void sendSms() {
+    }
+
+    public void sendMms() {
+    }
+
+    public void makePhoneCall(int seconds) {
+    }
```
Jeżeli to klasa abstrakcyjna, to najlepiej, żeby te metody też takie były

slawekludwiczak commented on src/main/java/pl/javastart/task/internal/Agreement.java:
```diff
@@ -0,0 +1,27 @@
+    int mmsCount = 0;
+    int secCount = 0;
+
+    public Agreement() {
```
niepotrzebnie jawnie zapisany konstruktor

slawekludwiczak commented on src/main/java/pl/javastart/task/internal/Subscription.java:
```diff
@@ -0,0 +1,36 @@
+
+public class Subscription extends Agreement {
+
+    public Subscription(double monthlyPayment) {
+        super();
+        this.monthlyPayment = monthlyPayment;
+    }
+
+    double monthlyPayment;
```
W kodzie zapisujemy najpierw pola, później konstruktory

slawekludwiczak commented on src/main/java/pl/javastart/task/internal/Prepaid.java:
```diff
@@ -0,0 +1,68 @@
+            smsCount++;
+            accountState -= smsPrice;
+            System.out.println("SMS wysłany\n");
+            return;
+        }
+        System.out.println("Nie udało się wysłać SMSa\n");
```
jeżeli metoda jest void, to lepiej użyć po prostu else zamiast returna

slawekludwiczak commented on src/main/java/pl/javastart/task/internal/Prepaid.java:
```diff
@@ -0,0 +1,68 @@
+            mmsCount++;
+            accountState -= mmsPrice;
+            System.out.println("MMS wysłany\n");
+            return;
+        }
+        System.out.println("Nie udało się wysłać MMSa\n");
```
jeżeli metoda jest void, to lepiej użyć po prostu else zamiast returna

slawekludwiczak commented on src/main/java/pl/javastart/task/internal/Mix.java:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    @Override
+    public void makePhoneCall(int seconds) {
+
+        if (secLimit <= 0 && accountState < (pricePerMin / 60)) {
+            System.out.println("Brak środków na koncie\n");
+            return;
+        }
+        int secondsFromLimit = 0;
+        if (secLimit > 0) {
+            secondsFromLimit = useSecondsFromLimit(seconds);
+        }
+        int secondsFromAccount = 0;
+        if (accountState > 0 && secondsFromLimit < seconds) {
+            secondsFromAccount = super.useSecondsFromAccount(seconds - secondsFromLimit);
+        }
+        int callDuration = secondsFromAccount + secondsFromLimit;
+        secCount += callDuration;
+        System.out.printf("Rozmowa trwała %d sekund\n\n", callDuration);
+    }
```
Może zostać tak, jak jest, ale tutaj warto by było wykorzystać logikę z klasy nadrzędnej, tzn.:
- jeżeli są darmowe sekundy to ich użyj,
- jeżeli rozmowa wciąż trwa, to wywołaj metodę z klasy Prepaid i spróbuj użyć środków na koncie (`super.makePhoneCall(remainingCallSeconds)`)
U Ciebie jest to o tyle kłopotliwe, że w klasach reprezentujących umowy masz System.out.printlny (w rzeczywistości miejscem interakcji z użytkownikiem powinien być raczej telefon, a nie umowa/karta sim)

jtermi replied:
wprowadziłem zmiany, przenosząc sout'y do klasy phone i zmieniając trochę logike metody. 
to jest cenna uwaga i warto było skorzystać, proszę nie tracić wiary ;)
jeśli coś jeszcze można usprawnić to bardzo będę wdzięczny za podpowiedź!

jtermi committed:
wprowadzone poprawki:
- zmieniłem metody w klasie Agreement na abstrakcyjne
- przeniosłem wszystkie wydruki do klasy Phone
- usunąłem konstruktor z klasy Agreement
- przenisłem definicje pola nad konstruktor (przypadkowo się tam znalazła, staram się pilnować kolejności)
- zamienilem return na else w 2ch metodach w klasie Prepaid.
- no i poprawiłem makePhoneCall w Mixie tak by można było lepiej skorzystać z agregacji, proszę nastęnym razem nie tracic wiary :) zgadzam się, że jest to ważna zmiana i warto było się tym zając!

jtermi committed:
wymagane zmiany wprowadziłem i zrozumiałem wreszcie do końca sens istnienia klasy phone :)
